### DISCUSSION TITLE
0052 XLS-52d: NFTokenMintOffer

### DISCUSSION THREAD

**Discussion by tequdev** [XRPLF/XRPL-Standards#147] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by mvadari** [comment]
Why edit the `NFTokenMint` transaction instead of creating a new transaction type?

**Reply by tequdev** [reply]
Considering the HookOn field in Hooks, I thought it would be a bad idea to unnecessarily increase the number of transaction types.
Is there any significant advantage to using this as a new transaction type?

**Reply by Ekiserrepe** [reply]
I think it's helpful to have a process that allows you to join both processes (it happens to me a lot) but as @mvadari says I was expecting to be a new tx type instead to modify the original one just for legacy issues for those who doesn't know the new change. I'm not against to rewrite the original tx but it's true it will create the issue to reeducate those who don't know the new changes. Thanks for the work @tequdev ❤️

**Reply by WietseWind** [reply]
> Considering the HookOn field in Hooks, I thought it would be a bad idea to unnecessarily increase the number of transaction types.

I agree with this. It's similar to flags imo, or even for example regular offers, where you can also cancel another offer optionally.

I would propose to prevent using more and more transaction types for things to be considered an option in an existing flow.

**Comment by shawnxie999** [comment]
Just curious why we need this change on the protocol level instead of calling `NFTokenMint` and `NFTokenCreateOffer` on client side? Just curious why you think calling these two separate transactions is problematic, leading to this proposal.

**Reply by tequdev** [reply]
This is especially useful for large random Mints rather than small creators.
If 1000 NFTs are used to create Mints and Offers at the same time, 2000 transactions are required.

Also, NFTokenCreateOffer requires an NFT ID that depends on MintedNFTokens, and the NFT ID required for the 1000th offer cannot be calculated in advance given the possibility of failure of any of the 1~999th NFTs to be minted.

It is more about improving convenience on the developer's side than on the user's side.

**Comment by mvadari** [comment]
Thoughts about this method in relation to XLS-56d? https://github.com/XRPLF/XRPL-Standards/discussions/162

**Reply by tequdev** [reply]
The idea is to combine NFTokenMint+NFTokenCreateOffer into a single Atomic transaction, but it still requires the NFT ID to be calculated.

It seems to me that both this and the Atomic feature can exist at the same time, and if so, the number of possible mints at the single Atomic transaction can go up to 8.

**Reply by ximinez** [reply]
Something to consider is that functionality always comes with costs, including potential for new bugs, maintenance over time, extra documentation, etc. If an `Atomic` transaction can do what `NFTokenMintOffer` can do (and it looks like it can), and the only difference is ~20 drops per mint and fewer mints per `Atomic`, then I don't think `NFTokenMintOffer` is worth the extra load and work.

**Reply by shawnxie999** [reply]
But NFTokenCreateOffer needs the NFT ID, which is generated only after the NFTokenMint is successful. Unless I'm missing something in the Atomic spec

**Reply by mvadari** [reply]
You could predict what it would be if it's successful.

**Reply by shawnxie999** [reply]
You could in most cases, but I would oppose doing that...

This is not robust and prone to changes on rippled

**Reply by xVet** [reply]
Yea we don't do that usually, calculating it ourselves, we take the NFTokenID from the metadata of the TX as this is actually what happens on chain.

**Comment by tequdev** [comment]
Since the implementation of this proposal has already been merged into the develop branch of rippled, this discussion will be closed and moved to the PR of this repository.

https://github.com/XRPLF/XRPL-Standards/pull/224
